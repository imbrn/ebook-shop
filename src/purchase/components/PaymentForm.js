import React, { PureComponent } from "react";
import styled from "styled-components";
import StepPanel, { StepPanelField, TextField, Button } from "./StepPanel";
import { Formik } from "formik";
import nanoid from "nanoid";

export const PaymentForm = ({
  open,
  number,
  title,
  payment,
  isValid,
  updatePayment,
  requestPurchase
}) => {
  const changePaymentMethod = method => {
    updatePayment({
      method
    });
  };

  return (
    <StepPanel open={open} number={number} title={title}>
      <div>
        <PaymentMethodButton
          selected={payment.method === "boleto"}
          onClick={() => changePaymentMethod("boleto")}
        >
          Boleto Banc√°rio
        </PaymentMethodButton>

        <PaymentMethodButton
          selected={payment.method === "creditCard"}
          onClick={() => changePaymentMethod("creditCard")}
        >
          Credit Card
        </PaymentMethodButton>
      </div>

      {payment.method === "creditCard" ? (
        <CreditCardForm
          payment={payment}
          isValid={isValid}
          updatePayment={updatePayment}
          requestPurchase={requestPurchase}
        />
      ) : (
        <BoletoForm isValid={isValid} requestPurchase={requestPurchase} />
      )}
    </StepPanel>
  );
};

PaymentForm.defaultProps = {
  open: true,
  number: 3,
  title: "Payment",
  payment: { method: "boleto" },
  isValid: false,
  updatePayment: () => {},
  requestPurchase: () => {}
};

const CreditCardForm = ({
  payment,
  isValid,
  updatePayment,
  requestPurchase
}) => {
  const cardholderNameFieldId = nanoid(8);
  const cardNumberFieldId = nanoid(8);
  const dueDateFieldId = nanoid(8);
  const cvvFieldId = nanoid(8);

  const onSubmit = e => {
    e.preventDefault();
    if (isValid) {
      requestPurchase();
    }
  };

  const handleInputChange = (e, handleChange) => {
    updatePayment({
      [e.target.name]: e.target.value
    });
    handleChange(e);
  };

  return (
    <Formik
      initialValues={{
        cardholderName: payment.cardholderName || "",
        cardNumber: payment.cardNumber || "",
        dueDate: payment.dueDate || "",
        cvv: payment.cvv || ""
      }}
      onSubmit={onSubmit}
      render={({
        values,
        touched,
        errors,
        handleChange,
        handleBlur,
        handleSubmit
      }) => (
        <form onSubmit={handleSubmit}>
          <StepPanelField
            component={TextField}
            fieldId={cardholderNameFieldId}
            name="cardholderName"
            label="Cardholder name"
            placeholder="Your name here"
            value={values.cardholderName}
            error={touched.cardholderName && errors.cardholderName}
            onChange={e => handleInputChange(e, handleChange)}
            onBlur={handleBlur}
          />

          <StepPanelField
            component={TextField}
            fieldId={cardNumberFieldId}
            name="cardNumber"
            label="Card number"
            placeholder="1234567891011121"
            value={values.cardNumber}
            error={touched.cardNumber && errors.cardNumber}
            onChange={e => handleInputChange(e, handleChange)}
            onBlur={handleBlur}
          />

          <StepPanelField
            component={TextField}
            fieldId={dueDateFieldId}
            name="dueDate"
            label="Due date"
            placeholder="12/25"
            value={values.dueDate}
            error={touched.dueDate && errors.dueDate}
            onChange={e => handleInputChange(e, handleChange)}
            onBlur={handleBlur}
          />

          <StepPanelField
            component={TextField}
            fieldId={cvvFieldId}
            name="cvv"
            label="CVV"
            placeholder="123"
            value={values.cvv}
            error={touched.cvv && errors.cvv}
            onChange={e => handleInputChange(e, handleChange)}
            onBlur={handleBlur}
          />

          <Button highlighted disabled={!isValid}>
            Buy now
          </Button>
        </form>
      )}
    />
  );
};

const BoletoForm = ({ isValid, requestPurchase }) => {
  const onSubmit = e => {
    e.preventDefault();
    requestPurchase();
  };

  return (
    <form onSubmit={onSubmit}>
      <Button highlighted disabled={!isValid}>
        Buy now
      </Button>
    </form>
  );
};

const PaymentMethodButton = styled(Button)``;
