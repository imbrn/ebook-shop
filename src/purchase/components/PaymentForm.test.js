import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import { PaymentForm } from "./PaymentForm";

jest.mock("nanoid", () => {
  let nextId = 1;
  return () => nextId++;
});

afterEach(cleanup);

test("renders correctly", () => {
  const payment = {
    method: "creditCard",
    cardholderName: "Customer Name",
    cardNumber: "4123456789012",
    dueDate: new Date(2025, 11),
    cvv: "123"
  };
  const { container } = render(<PaymentForm payment={payment} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("updatePayment", () => {
  const updatePayment = jest.fn();

  const { getByLabelText } = render(
    <PaymentForm
      payment={{ method: "creditCard" }}
      updatePayment={updatePayment}
    />
  );

  fireEvent.change(getByLabelText("Cardholder name"), {
    target: { value: "Customer Name" }
  });

  fireEvent.change(getByLabelText("Card number"), {
    target: { value: "4123456789012" }
  });

  fireEvent.change(getByLabelText("Due date"), {
    target: { value: "12/25" }
  });

  fireEvent.change(getByLabelText("CVV"), {
    target: { value: "123" }
  });

  expect(updatePayment.mock.calls).toEqual([
    [{ cardholderName: "Customer Name" }],
    [{ cardNumber: "4123456789012" }],
    [{ dueDate: "12/25" }],
    [{ cvv: "123" }]
  ]);
});

test("requestPurchase", () => {
  const requestPurchase = jest.fn();
  const { getByText } = render(
    <PaymentForm isValid={true} requestPurchase={requestPurchase} />
  );
  fireEvent.click(getByText("Buy now"));
  expect(requestPurchase.mock.calls).toHaveLength(1);
});
