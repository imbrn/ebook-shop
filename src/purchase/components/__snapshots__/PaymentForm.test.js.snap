// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <header>
    <span
      class=""
    >
      3
    </span>
    <span
      class=""
    >
      Payment
    </span>
  </header>
  <div>
    <div>
      <button
        class=""
      >
        Boleto Banc√°rio
      </button>
      <button
        class=""
      >
        Credit Card
      </button>
    </div>
    <form>
      <div>
        <div>
          <label
            class=""
            for="1"
          >
            Cardholder name
          </label>
        </div>
        <div>
          <input
            class=""
            id="1"
            name="cardholderName"
            placeholder="Your name here"
            type="text"
            value="Customer Name"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="2"
          >
            Card number
          </label>
        </div>
        <div>
          <input
            class=""
            id="2"
            name="cardNumber"
            placeholder="1234567891011121"
            type="text"
            value="4123456789012"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="3"
          >
            Due date
          </label>
        </div>
        <div>
          <input
            class=""
            id="3"
            name="dueDate"
            placeholder="12/25"
            type="text"
            value="Mon Dec 01 2025 00:00:00 GMT-0200 (-02)"
          />
        </div>
      </div>
      <div>
        <div>
          <label
            class=""
            for="4"
          >
            CVV
          </label>
        </div>
        <div>
          <input
            class=""
            id="4"
            name="cvv"
            placeholder="123"
            type="text"
            value="123"
          />
        </div>
      </div>
      <button
        class=""
        disabled=""
      >
        Buy now
      </button>
    </form>
  </div>
</div>
`;
